### PR TITLE
Download new needles from openQA on resume

### PR DESCRIPTION
**Pull request by Martchus**
see https://progress.opensuse.org/issues/41075
openQA: https://github.com/os-autoinst/openQA/pull/1793

### CODE REVIEW

**Comment by Martchus** [comment]
BTW, this is supposed to work like this:

1. On resume, the liveviewhandler checks for needles which have been created/updated since the running test has been started. That list is attached to the resume command passt to os-autoinst.
2. os-autoinst will go through the list and download missing/outdated needles before resuming. The download is done in isotovideo. The download is done though new web API routes (*not* served by liveviewhandler).

This approach seems to work in staging where caching is enabled: https://e212.suse.de/tests/3

So I'll keep polishing it. TODOs:

* Fix warnings
* Modularize download and add unit tests
* Prevent downloading SLE needles for an openSUSE test
* Prevent downloading non-existing needles

File: OpenQA/Isotovideo/NeedleDownloader.pm [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+use Mojo::File;
+use File::stat;
+use Try::Tiny;
+use POSIX 'strftime';
+use testapi 'diag';
+use bmwqemu;
+
+has files_to_download => sub { [] };
+has needle_dir        => sub { needle::default_needle_dir() };
+has openqa_url        => sub { 'http://' . $bmwqemu::vars{OPENQA_URL} };
```
Comment (okurz) [inline OpenQA/Isotovideo/NeedleDownloader.pm]: As I understood the variable "OPENQA_URL" normally includes the prefix, e.g. see https://openqa.opensuse.org/tests/758251/file/vars.json . Maybe better use an URL parsing function which I am sure we have used somewhere else already.

File: OpenQA/Isotovideo/NeedleDownloader.pm [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    my $needle_name     = $needle->{name};
+    my $latest_update   = $needle->{t_updated};
+    my $needle_dir      = $self->needle_dir;
+    my $download_target = "$needle_dir/$needle_name.$extension";
+
+    if (my $target_stat = stat($download_target)) {
+        if (my $target_last_modified = $target_stat->[9] // $target_stat->[8]) {
+            $target_last_modified = strftime('%Y-%m-%dT%H:%M:%SZ', gmtime($target_last_modified));
+            if ($target_last_modified ge $latest_update) {
+                diag("skip downloading new needle: $download_target seems already up-to-date (last update: $target_last_modified > $latest_update)");
```
Comment (okurz) [inline OpenQA/Isotovideo/NeedleDownloader.pm]: I see imperative tense in logging messages as misleading. -> s/skip/Skipping/

File: OpenQA/Isotovideo/NeedleDownloader.pm (hunk 1 above)
Comment (Martchus) [reply OpenQA/Isotovideo/NeedleDownloader.pm]: Ok - than the name of the variable would actually make sense. Strange that in my case the 'URL' was actually just the hostname. I'll improve that to handle both cases. We usually use `Mojo::URL` for this.

File: OpenQA/Isotovideo/NeedleDownloader.pm (hunk 2 above)
Comment (Martchus) [reply OpenQA/Isotovideo/NeedleDownloader.pm]: Right, I'll change that.

**Comment by Martchus** [comment]
Not sure how to make it ignore unneeded needles with reasonable effort. But just downloading any new or updated needle since the test has been started is not too wasteful, isn't it?

**Comment by coolo** [comment]
I don't think so. Just add a sanity max - to avoid something going busted, e.g. at max 100

**Comment by Martchus** [comment]
Ok, that should be simple enough.

I also suppose it isn't the best idea to do the blocking downloads in isotovideo which will be unresponsive in that time. Do you the download is better done in autotest?

**Comment by coolo** [comment]
autotest is blocked in check_screen - hooking a download into that sounds like a spaghetti code nightmare. Possibly easiest is really a temporary child process of isotovideo. But that's just brain dump - do your own experiments.

**Comment by Martchus** [comment]
> autotest is blocked in check_screen - hooking a download into that sounds like a spaghetti code nightmare.

I just checked out the code again. I suppose moving the download to autotest would be as simple as moving the download [before the reload is triggered](https://github.com/os-autoinst/os-autoinst/blob/master/testapi.pm#L340). Only disadvantage is that it wouldn't work if the test is paused for another reason than check/assert screen timeout. (Btw, in this case the needle reload isn't triggered so far.)

**Comment by Martchus** [comment]
After adding limit and doing the download in autotest it still works on staging: https://e212.suse.de/tests/9

**Comment by Martchus** [comment]
Rebased and fixed perlcritic error. Let's see whether the tests pass now.

**Comment by okurz** [comment]
Broke the build on all repos in [devel:openQA](https://build.opensuse.org/package/show/devel:openQA/os-autoinst)

* https://build.opensuse.org/request/show/640378 adds the new dependency to os-autoinst.spec
* https://build.opensuse.org/request/show/640381 and https://build.opensuse.org/request/show/640382 adds perl-File-Touch to the according SLE subprojects of devel:openQA

**Pull request merged**